src/backend/columnar/columnar_tableam.c

```diff
@@ -1372,21 +1374,145 @@ ColumnarReadRowsIntoIndex(TableScanDesc scan, Relation indexRelation,
+											greatestIndexItemPointer, estate,
+											predicate, validateIndexState);
+
+	table_endscan(scan);
+
+	/* report the last "virtual" block as "done" */
+	ColumnarReportTotalVirtualBlocks(columnarRelation, snapshot,
+									 PROGRESS_SCAN_BLOCKS_DONE);
+
+	ExecDropSingleTupleTableSlot(econtext->ecxt_scantuple);
+	FreeExecutorState(estate);
+	indexInfo->ii_ExpressionsState = NIL;
+	indexInfo->ii_PredicateState = NULL;
+}
+
+
+/*
+ * FindGreatestIndexItemPointer returns ItemPointerData for the index tuple
+ * with greatest tid.
+ * If index is empty, then returns invalid ItemPointerData.
+ */
+static ItemPointerData
+FindGreatestIndexItemPointer(Relation relation, Relation indexRelation, Snapshot snapshot)
+{
+	ItemPointerData greatestIndexItemPointer;
+	ItemPointerSetInvalid(&greatestIndexItemPointer);
+
+	int nkeys = 0;
+	int norderbys = 0;
+	IndexScanDesc indexScan = index_beginscan(relation, indexRelation,
+											  snapshot, nkeys, norderbys);
+	ItemPointer indexItemPointer = NULL;
+	while ((indexItemPointer = index_getnext_tid(indexScan, ForwardScanDirection)))
+	{
+		if (!ItemPointerIsValid(&greatestIndexItemPointer) ||
+			ItemPointerCompare(indexItemPointer, &greatestIndexItemPointer) > 0)
+		{
+			greatestIndexItemPointer = *indexItemPointer;
+		}
+	}
+	index_endscan(indexScan);
+
+	return greatestIndexItemPointer;
+}
+
+
+/*
+ * ColumnarReadRowsWithGreaterTidIntoIndex inserts the tuples that are not in
+ * the index yet by reading the actual relation based on given "scan".
+ */
+static void
+ColumnarReadRowsWithGreaterTidIntoIndex(TableScanDesc scan, Relation indexRelation,
+										IndexInfo *indexInfo,
+										ItemPointerData greatestIndexItemPointer,
+										EState *estate, ExprState *predicate,
+										ValidateIndexState *state)
+{
+	BlockNumber lastReportedBlockNumber = InvalidBlockNumber;
+
+	ExprContext *econtext = GetPerTupleExprContext(estate);
+	TupleTableSlot *slot = econtext->ecxt_scantuple;
+	while (columnar_getnextslot(scan, ForwardScanDirection, slot))
+	{
+		CHECK_FOR_INTERRUPTS();
+
+		BlockNumber currentBlockNumber = ItemPointerGetBlockNumber(&slot->tts_tid);
+		if (lastReportedBlockNumber != currentBlockNumber)
+		{
+			/*
+			 * columnar_getnextslot guarantees that returned tuple will
+			 * always have a greater ItemPointer than the ones we fetched
+			 * before, so we directly use BlockNumber to report our progress.
+			 */
+			Assert(lastReportedBlockNumber == InvalidBlockNumber ||
+				   currentBlockNumber >= lastReportedBlockNumber);
+			pgstat_progress_update_param(PROGRESS_SCAN_BLOCKS_DONE,
+										 currentBlockNumber);
+			lastReportedBlockNumber = currentBlockNumber;
+		}
+
+		MemoryContextReset(econtext->ecxt_per_tuple_memory);
+
+		state->htups += 1;
+
+		if (ItemPointerIsValid(&greatestIndexItemPointer) &&
+			ItemPointerCompare(&slot->tts_tid, &greatestIndexItemPointer) <= 0)
+		{
+			/* tuple is already covered by the index, skip */
+			continue;
+		}
```

Comment:
Here we use the highest already-indexed `ItemPointer` to skip the tuples that are already covered by the index.

It would be very nice if we can push this logic down to `columnar_getnextslot` so that we can directly skip the stripes that are all covered by the index.

But I think it's too much work so we can do this later.

Reply:
This already wasn't the correct approach, so skipping.
https://github.com/citusdata/citus/pull/5032#discussion_r654318829